flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java

```diff
@@ -560,53 +621,177 @@ public ExecutionContext getExecutionContext() {
 	}
 
 	/**
+	 * This works as cache for already merged accumulators, as, in some cases,
+	 * we do not want to remerge accumulators as this may lead to duplicate entries.
+	 */
+	private Map<String, Accumulator<?, ?>> mergedSmallUserAccumulators;
```

Comment:
Why can it lead to duplicate entries? The accumulators should always be merged from scratch. Every accumulator update message contains the current state of the accumulators.